mlir/lib/Dialect/Rock/Transforms/Regularize.cpp

```diff
@@ -147,187 +146,452 @@ struct RegularizeGenericRewritePattern
+  std::optional<SmallVector<OpOperand *>> writers =
+      bufferDeps.getWriters(*maybeBuffer);
+  if (!writers)
+    return;
+  for (OpOperand *writer : *writers) {
+    if (!state.inputFusionWrites.insert(writer).second) {
+      continue; // No infinite recursion
     }
-    chain.push_back(forwOp);
-    if (auto lgop = dyn_cast<linalg::GenericOp>(forwOp)) {
-      return llvm::is_contained(lgop.getOutputs(), result);
-    } else if (auto mcop = dyn_cast<memref::CopyOp>(forwOp)) {
-      // should never be output of memcpy?
-      assert(mcop.getTarget() != result);
-      return mcop.getTarget() == result;
-    } else if (auto rgop = dyn_cast<rock::GridwiseGemmOp>(forwOp)) {
-      return rgop.getC() == result;
-    } else if (auto rgop = dyn_cast<rock::GridwiseGemmAccelOp>(forwOp)) {
-      return rgop.getC() == result;
-    } else if (auto rgop = dyn_cast<rock::GridwiseAttentionAccelOp>(forwOp)) {
-      return rgop.getOut() == result;
-    } else if (auto reduceOp = dyn_cast<rock::ReduceOp>(forwOp)) {
-      return reduceOp.getOut() == result;
-    } else if (auto rgop = dyn_cast<rock::ThreadwiseWriteAllOp>(forwOp)) {
-      return rgop.getDest() == result;
+    auto writeOp = dyn_cast<MemoryEffectOpInterface>(writer->getOwner());
+    if (!writeOp)
+      continue;
+    SmallVector<MemoryEffects::EffectInstance> effects;
+    writeOp.getEffects(effects);
+    for (const auto &effect : effects) {
+      OpOperand *maybeRecursiveReader = effect.getEffectValue<OpOperand *>();
+      // Test against `writer` to guard against [MemRead, MemWrite]
+      if (maybeRecursiveReader && maybeRecursiveReader != writer &&
+          isa<MemoryEffects::Read>(effect.getEffect())) {
+        collectInputFusionWrites(maybeRecursiveReader, bufferDeps, state);
+      }
     }
-    LLVM_DEBUG(llvm::dbgs() << "unsupported op\n" << *forwOp);
-    return false;
   }
+}
 
-  LogicalResult matchAndRewrite(memref::AllocOp alloc,
-                                PatternRewriter &rw) const override {
-    LogicalResult lres = failure();
-    Value buffer = alloc.getResult();
-
-    bool hasTransforms = false;
-    Operation *writer = nullptr;
-    Operation *fusor = nullptr;
-    SmallVector<SmallVector<Operation *>> readChains;
-
-    // find the fusor
-    for (auto &use : buffer.getUses()) {
-      Operation *useOp = use.getOwner();
-      Value result = buffer;
-      while (auto top = dyn_cast<rock::TransformOp>(useOp)) {
-        result = top.getResult();
-        useOp = (*result.getUses().begin()).getOwner();
+/// Given the operand that writes to a buffer, collect the writes of its reader,
+/// and the writes of that reader, and so on until you're no longer writing to a
+/// buffer. If any auxiliary inputs come from buffers themselves, collect
+/// their writes as input fusions, because, in a fusion context, those other
+/// writers will be tiled by a threadwise_read_into, not a threadwise_write_all.
+static void
+collectOutputFusionWrites(OpOperand *writer,
```

Comment:
Same confusion here around how an OpOperand be a writer... what am I missing ?

Reply:
The `writer` is, for example, the `c` argument on a GEMM or the output of a linalg.generic